README.md
# domoticz_linky
Get Linky smart meter data and push it to domoticz

If you appreciate this software, please show it off ! [![PayPal donate button](http://img.shields.io/paypal/donate.png?color=yellow)](https://www.paypal.com/cgi-bin/webscr?cmd=_xclick&business=epierre@e-nef.com&currency_code=EUR&amount=&item_name=thanks "Donate once-off to this project using Paypal")


# pre-requisite: activate load curve recording on Enedis website
enable your enedis account (https://espace-client-particuliers.enedis.fr/group/espace-particuliers/) and the data collection ("Consommation" > "Gérer ma courbe de charge" > "Activer ma courbe de charge")

# create a device in Domoticz
- In Domoticz, go to hardware, create a virtual "rfx meter counter" or "Dummy".
- Then in Devices, add it to the devices. (mark down the id for later).
- When in Utility, edit the device and change it to Electric (instant+counter)  type .

## modules to install

    sudo apt-get install sqlite3 node npm
    sudo apt-get install python3 python3-numpy python3-dateutil python3-requests
## deploy
    cd <whereyouwant> ie. /home/pi/domoticz/plugins/
    git clone https://github.com/empierre/domoticz_linky.git
    cd domoticz_linky
    npm install winston

## rename configuration file, change login/pass/id

    cp _domoticz_linky.cfg domoticz_linky.cfg
    nano domoticz_linky.cfg

and change:

    LINKY_USERNAME="username@mail.com"
    LINKY_PASSWORD="password"
    DOMOTICZ_ID=547
    
Where DOMOTICZ_ID is id device on domoticz. 

Configuration file will not be deleted in future updates


## testing before launch

Manually launch

    ./domoticz_linky.sh

N.B. If login is not ok, you'll get a nodejs error on console for data will be missing (will be changed).

Then check the login credential if they are ok:

    domoticz_linky.log

If this is good, you'll get several json files in the directory

## Add to your cron tab (with crontab -e):

    30 7,17 * * * /home/pi/domoticz_linky/domoticz_linky.sh

## Getting back history

  It is possible to get back history months giving the relative month to get. Parameter is how many month back from today. Limitation: works only on same year at this time.

    ./domoticz_linky_month.sh 3
    
 On example, 3 months before.


CHANGELOG
v20180318 - arigaud
  - unique configuration file
  - unique script to daily/monthly to gets datas
  - added error when website is on maintenance
  - install domoticz_linky directory where you want


linky.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Retrieves energy consumption data from your Enedis (ERDF) account.
"""
# Linkindle - Linky energy consumption curves on a Kindle display.
# Copyright (C) 2016 Baptiste Candellier
# Modified (C) 2017 Asdepique777
# corrected (C) 2017 epierre
# Modified (C) 2017 Asdepique777
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import base64
import requests
import html
import sys
import os
import logging
import json

LOGIN_BASE_URI = 'https://espace-client-connexion.enedis.fr'
API_BASE_URI = 'https://espace-client-particuliers.enedis.fr/group/espace-particuliers'

API_ENDPOINT_LOGIN = '/auth/UI/Login'
API_ENDPOINT_HOME = '/accueil'
API_ENDPOINT_DATA = '/suivi-de-consommation'

DATA_NOT_REQUESTED = -1
DATA_NOT_AVAILABLE = -2

class LinkyLoginException(Exception):
    """Thrown if an error was encountered while retrieving energy consumption data."""
    pass

class LinkyServiceException(Exception):
    """Thrown when the webservice threw an exception."""
    pass

def login(username, password):
    """Logs the user into the Linky API.
    """
    session = requests.Session()

    payload = {'IDToken1': username,
               'IDToken2': password,
               'SunQueryParamsString': base64.b64encode(b'realm=particuliers'),
               'encoded': 'true',
               'gx_charset': 'UTF-8'}

    session.headers = {
                'User-Agent': 'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Mobile Safari/537.36',
                'Accept-Language':'fr,fr-FR;q=0.8,en;q=0.6',
                'Accept-Encoding':'gzip, deflate',
                'Accept':'application/json, text/javascript, */*; q=0.01'}


    req = session.post(LOGIN_BASE_URI + API_ENDPOINT_LOGIN, data=payload, allow_redirects=False)

    session_cookie = req.cookies.get('iPlanetDirectoryPro')

    if not 'iPlanetDirectoryPro' in session.cookies:
        raise LinkyLoginException("Login unsuccessful. Check your credentials.")

    return session


def get_data_per_hour(session, start_date, end_date):
    """Retreives hourly energy consumption data."""
    return _get_data(session, 'urlCdcHeure', start_date, end_date)

def get_data_per_day(session, start_date, end_date):
    """Retreives daily energy consumption data."""
    return _get_data(session, 'urlCdcJour', start_date, end_date)

def get_data_per_month(session, start_date, end_date):
    """Retreives monthly energy consumption data."""
    return _get_data(session, 'urlCdcMois', start_date, end_date)

def get_data_per_year(session):
    """Retreives yearly energy consumption data."""
    return _get_data(session, 'urlCdcAn')

def _get_data(session, resource_id, start_date=None, end_date=None):
    req_part = 'lincspartdisplaycdc_WAR_lincspartcdcportlet'


    # We send the session token so that the server knows who we are
    payload = {
        '_' + req_part + '_dateDebut': start_date,
        '_' + req_part + '_dateFin': end_date
    }

    params = {
        'p_p_id': req_part,
        'p_p_lifecycle': 2,
        'p_p_state': 'normal',
        'p_p_mode': 'view',
        'p_p_resource_id': resource_id,
        'p_p_cacheability': 'cacheLevelPage',
        'p_p_col_id': 'column-1',
        'p_p_col_pos': 1,
        'p_p_col_count': 3
    }

    req = session.post(API_BASE_URI + API_ENDPOINT_DATA, allow_redirects=False, data=payload, params=params)

    if 300 <= req.status_code < 400:
        # So... apparently, we may need to do that once again if we hit a 302
        # ¯\_(ツ)_/¯
        req = session.post(API_BASE_URI + API_ENDPOINT_DATA, allow_redirects=False, data=payload, params=params)

    if req.status_code == 200 and req.text is not None and "Conditions d'utilisation" in req.text:
        raise LinkyLoginException("You need to accept the latest Terms of Use. Please manually log into the website, "
                                  "then come back.")
    if req.status_code == 200 and req.text is not None and "Une erreur technique" in req.text:
        raise LinkyLoginException("A technical error has occurred on website. Data unavailable.")

    try:
        res = json.loads(req.text)
    except:
        logging.info("Unable to log in")
        sys.exit(os.EX_SOFTWARE)

    if res['etat'] and res['etat']['valeur'] == 'erreur' and res['etat']['erreurText']:
        raise LinkyServiceException(html.unescape(res['etat']['erreurText']))

    return res


linky_month.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
"""Generates energy consumption JSON files from Enedis (ERDF) consumption data
collected via their  website (API).
"""

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import datetime
import logging
import sys
import json
import linky
from dateutil.relativedelta import relativedelta

USERNAME = os.environ['LINKY_USERNAME']
PASSWORD = os.environ['LINKY_PASSWORD']
BASEDIR = os.environ['BASE_DIR']

# Generate y axis (consumption values) 
def generate_y_axis(res):
    y_values = []

    # Extract data points from the source dictionary into a list
    for ordre, datapoint in enumerate(res['graphe']['data']):
        value = datapoint['valeur']

        # Remove any invalid values
        # (they're error codes on the API side, but useless here)
        if value < 0:
            value = 0

        y_values.insert(ordre, value)

    return y_values

# Generate x axis (time values)  
def generate_x_axis(res, time_delta_unit, time_format, inc):
    x_values = []

    # Extract start date and parse it
    start_date_queried_str = res['graphe']['periode']['dateDebut']
    start_date_queried = datetime.datetime.strptime(start_date_queried_str, "%d/%m/%Y").date()

    # Calculate final start date using the "offset" attribute returned by the API
    kwargs = {}
    kwargs[time_delta_unit] = res['graphe']['decalage'] * inc
    start_date = start_date_queried - relativedelta(**kwargs)

    # Generate X axis time labels for every data point
    for ordre, _ in enumerate(res['graphe']['data']):
        kwargs = {}
        kwargs[time_delta_unit] = ordre * inc
        x_values.insert(ordre, (start_date + relativedelta(**kwargs)).strftime(time_format))

    return x_values

# Date formatting 
def dtostr(date):
    return date.strftime("%d/%m/%Y")


# Export the JSON file for half-hours power measure (for the last pas day)
def export_hours_values(res):
    hours_x_values = generate_x_axis(res, \
                                    'hours', "%H:%M", 0.5)
    hours_y_values = generate_y_axis(res)
    hours_values = []

    for i in range(0,len(hours_x_values)):
        hours_values.append({"time" : hours_x_values[i], "conso" : hours_y_values[i]})
    with open(BASEDIR+"/export_hours_values.json", 'w+') as outfile:
        json.dump(hours_values, outfile)

# Export the JSON file for daily consumption (for the past rolling 30 days)
def export_days_values(res):
    days_x_values = generate_x_axis(res, \
                                    'days', "%d %b", 1)
    days_y_values = generate_y_axis(res)
    days_values = []

    for i in range(0,len(days_x_values)):
        days_values.append({"time" : days_x_values[i], "conso" : days_y_values[i]})
    with open(BASEDIR+"/export_days_values.json", 'w+') as outfile:
        json.dump(days_values, outfile)

# Export the JSON file for monthly consumption (for the current year, starting 12 months from today)
def export_months_values(res):
    months_x_values = generate_x_axis(res, \
                                    'months', "%b", 1)
    months_y_values = generate_y_axis(res)
    months_values = []

    for i in range(0,len(months_x_values)):
        months_values.append({"time" : months_x_values[i], "conso" : months_y_values[i]})
    with open(BASEDIR+"/export_months_values.json", 'w+') as outfile:
        json.dump(months_values, outfile)

# Export the JSON file for yearly consumption
def export_years_values(res):
    years_x_values = generate_x_axis(res, \
                                    'years', "%Y", 1)
    years_y_values = generate_y_axis(res)
    years_values = []

    for i in range(0,len(years_x_values)):
        years_values.append({"time" : years_x_values[i], "conso" : years_y_values[i]})
    with open(BASEDIR+"/export_years_values.json", 'w+') as outfile:
        json.dump(years_values, outfile)




# Main script 
def main():
    logging.basicConfig(format='%(asctime)s %(message)s', level=logging.INFO)

    try:
        logging.info("logging in as %s...", USERNAME)
        token = linky.login(USERNAME, PASSWORD)
        logging.info("logged in successfully!")

        logging.info("retreiving data...")
        today = datetime.date.today()
        logging.info("arg "+sys.argv[1])
        mmonth     = int(sys.argv[1])
        mmonthnext = int(sys.argv[1])+1
        logging.info(dtostr(today - relativedelta(days=1, months=mmonth)))
        logging.info(dtostr(today - relativedelta(days=1, months=mmonthnext)))
        

        # 12 months ago - today
        res_month = linky.get_data_per_month(token, dtostr(today - relativedelta(months=11)), \
                                             dtostr(today))


        # One month ago - yesterday
        res_day = linky.get_data_per_day(token, dtostr(today - relativedelta(days=1, months=mmonthnext)), \
                                         dtostr(today - relativedelta(days=1, months=mmonth)))

        

        logging.info("got data!")
############################################
		# Export of the JSON files, with exception handling as Enedis website is not robust and return empty data often

        try:
            export_days_values(res_day)
        except Exception:
            logging.info("days values non exported")


############################################
 
    except linky.LinkyLoginException as exc:
        logging.error(exc)
        sys.exit(1)



if __name__ == "__main__":
    main()


domoticz_linky.js
//##############################################################################
//  This file is part of domoticz_linky - https://github.com/empierre/domotics_linky
//      Copyright (C) 2014-2019 Emmanuel PIERRE (domoticz@e-nef.com)
//
//  domoticz_linky is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 2 of the License, or
//  (at your option) any later version.
//
//  MyDomoAtHome is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with MyDomoAtHome.  If not, see <http://www.gnu.org/licenses/>.
//##############################################################################

var fs = require('fs');
var winston = require('winston');
global.logger = winston;

const path = require('path');

var devicerowid=process.argv[2];
var dateObj = new Date();
var q_year=dateObj.getUTCFullYear();
var q_month_s=dateObj.getUTCMonth();
var q_month_e=dateObj.getUTCMonth() + 1;
var q_day_s=dateObj.getUTCDate()-1;
var q_day_e=dateObj.getUTCDate();

var BASE_DIR = process.env.BASE_DIR || '/home/pi/domoticz/domoticz_linky';

function pad(n, width, z) {
  z = z || '0';
  n = n + '';
  return n.length >= width ? n : new Array(width - n.length + 1).join(z) + n;
}

function getTotal() {
        try {
                var fileExport = 'export_years_values.json';
                var filePath = path.resolve(BASE_DIR, fileExport);
                var obj = JSON.parse(fs.readFileSync(filePath, 'utf8'));

                var conso_cumul=0.0;
                for (var i = 0; i < Object.keys(obj).length; ++i) {
                         conso_cumul= conso_cumul+ (obj[i]["conso"]);
                }
                return(conso_cumul);
        } catch (e) {
                // It isn't accessible
                console.log("Exception opening export_years_values.json : "+e);
        }

}
function getYear(year) {
        try {
                var fileExport = 'export_years_values.json';
                var filePath = path.resolve(BASE_DIR, fileExport);
                var obj = JSON.parse(fs.readFileSync(filePath, 'utf8'));
                for (var i = 0; i < Object.keys(obj).length; ++i) {
                         if (year == obj[i]["time"]) return (obj[i]["conso"]);
                }
        } catch (e) {
                // It isn't accessible
                console.log("Exception opening export_years_values.json : "+e);
        }
}
function getMonth(month) {
        var mth=[ 'Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'];
        try {
                var fileExport = 'export_months_values.json';
                var filePath = path.resolve(BASE_DIR, fileExport);
                var obj = JSON.parse(fs.readFileSync(filePath, 'utf8'));

                for (var i = 0; i < Object.keys(obj).length; ++i) {
                         if (mth[month-1] == obj[i]["time"]) return (obj[i]["conso"]);
                }
        } catch (e) {
                // It isn't accessible
                console.log("Exception opening export_months_values.json : "+e);
        }
}
function getDay(day,month) {
        var mth=[ 'Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'];
        try {
                var fileExport = 'export_days_values.json';
                var filePath = path.resolve(BASE_DIR, fileExport);
                var obj = JSON.parse(fs.readFileSync(filePath, 'utf8'));

                var dm=''+pad(day,2)+" "+mth[month-1];
                for (var i = 0; i < Object.keys(obj).length; ++i) {
                         if (dm == obj[i]["time"]) return (obj[i]["conso"]);
                }
        } catch (e) {
                // It isn't accessible
                console.log("Exception opening export_days_values.json : "+e);
        }
}
function generateDayHours() {
        var myDateObj = new Date();
        myDateObj.setUTCSeconds(0);
        myDateObj.setUTCMinutes(0);
        myDateObj.setUTCHours(0);
        var dateOffset = (24*60*60*1000) * 2; //2 days
        var dateOffset2 = 60*60*1000; //60 min
        myDateObj.setTime(myDateObj.getTime() - dateOffset);
        var cumul=getCumulBefore(myDateObj.getUTCFullYear(),myDateObj.getUTCMonth());
        try {
                var fileExport = 'export_hours_values.json';
                var filePath = path.resolve(BASE_DIR, fileExport);
                var obj = JSON.parse(fs.readFileSync(filePath, 'utf8'));

                console.log('BEGIN TRANSACTION;') ;
                console.log('DELETE FROM \'Meter\' WHERE devicerowid='+devicerowid+';') ;
                for (var i = 0; i+1 < Object.keys(obj).length; i=i+2) {
                        myDateObj.setTime(myDateObj.getTime() + dateOffset2);
                        var req_date= myDateObj.getUTCFullYear() + "-" + ("0"+(myDateObj.getUTCMonth()+1)).slice(-2) + "-" + ("0" + myDateObj.getUTCDate()).slice(-2) + " " + ("0" + myDateObj.getUTCHours()).slice(-2) + ":" + ("0" + myDateObj.getUTCMinutes()).slice(-2) + ":00";
                        var conso = (obj[i]["conso"] + obj[i+1]["conso"]) / 2;
                        if (conso > 0) {
                                console.log('INSERT INTO \'Meter\' (DeviceRowID,Usage,Value,Date) VALUES ('+devicerowid+', \''+Math.round(conso*10000/2)+'\', \''+Math.round(cumul*1000)+'\', \''+req_date+'\');') ;
                                cumul=cumul+conso;
                        }
                }
                myDateObj.setTime(myDateObj.getTime() + dateOffset2);
                console.log('INSERT INTO \'Meter\' (DeviceRowID,Usage,Value,Date) VALUES ('+devicerowid+', \''+0+'\', \''+Math.round(cumul*1000)+'\', \''+req_date+'\');') ;
                console.log('COMMIT;') ;
        } catch (e) {
                // It isn't accessible
                console.log('ROLLBACK;') ;
                console.log("Exception opening export_hours_values.json : "+e);
        }
}
function getCumulBefore(year,month) {
        // Bring back the year-month previous total as domoticz expect it
        try {
                var fileExport = 'export_years_values.json';
                var filePath = path.resolve(BASE_DIR, fileExport);
                var obj = JSON.parse(fs.readFileSync(filePath, 'utf8'));

                var conso_cumul=0.0;
                for (var i = 0; i < Object.keys(obj).length; ++i) {
                         if (obj[i]["time"]<year) {
                                 conso_cumul= conso_cumul+ (obj[i]["conso"]);
                        }
                }
        } catch (e) {
                // It isn't accessible
                console.log("Exception opening export_years_values.json : "+e);
        }
        var mth=[ 'Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'];
        try {
                var fileExport = 'export_months_values.json';
                var filePath = path.resolve(BASE_DIR, fileExport);
                var obj = JSON.parse(fs.readFileSync(filePath, 'utf8'));
                for (var i = 0; i < month-1; ++i) {
                         conso_cumul= conso_cumul+ (obj[i]["conso"]);
                }
                return(conso_cumul);
        } catch (e) {
                // It isn't accessible
                console.log("Exception opening export_months_values.json : "+e);
        }
}
function generateMonthDays() {
        var cumul=getCumulBefore(q_year,q_month_s);
        var mth=[ 'Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'];

        try {
                var fileExport = 'export_days_values.json';
                var filePath = path.resolve(BASE_DIR, fileExport);
                var obj = JSON.parse(fs.readFileSync(filePath, 'utf8'));
                var lastVal = 0;
                console.log('BEGIN TRANSACTION;') ;
                for (var i = 0; i < Object.keys(obj).length; ++i) {
                        var req_date=''+q_year+'-'+pad((mth.indexOf(obj[i]["time"].substr(3, 3))+1),2)+'-'+pad(obj[i]["time"].substr(0, 2),2);
                        if (obj[i]["conso"]>0) {
                                lastVal = Number((obj[i]["conso"]*1000).toFixed(2));
                                console.log('DELETE FROM \'Meter_Calendar\' WHERE devicerowid='+devicerowid+' and date = \''+req_date+'\'; INSERT INTO \'Meter_Calendar\' (DeviceRowID,Value,Counter,Date) VALUES ('+devicerowid+', \''+lastVal+'\', \''+Math.round(cumul*1000)/1000+'\', \''+req_date+'\');') ;
                                cumul=cumul+(obj[i]["conso"]);
                        }
                }
                if (i > 0) {
                    var myDateObj = new Date();
                    var dateOffset = (24*60*60*1000) * 1; //1 day
                    myDateObj.setTime(myDateObj.getTime() + dateOffset);
                    var req_date= myDateObj.getUTCFullYear() + "-" + ("0"+(myDateObj.getUTCMonth()+1)).slice(-2) + "-" + ("0" + myDateObj.getUTCDate()).slice(-2) + " " + ("0" + myDateObj.getUTCHours()).slice(-2) + ":" + ("0" + myDateObj.getUTCMinutes()).slice(-2) + ":00";
                    //  prevent device to go red
                    // Doesn't work well
                    // console.log('UPDATE DeviceStatus SET LastUpdate=\''+req_date+'\' WHERE ID==\''+ devicerowid+'\';');              
                    // console.log('UPDATE DeviceStatus SET sValue=\''+lastVal+'\', LastUpdate=\''+req_date+'\' WHERE ID==\''+ devicerowid+'\';');              
                }
                console.log('COMMIT;') ;
        } catch (e) {
                // It isn't accessible
                console.log('ROLLBACK;') ;
                console.log("Exception opening export_months_values.json : "+e);
        }
}


logger.add(winston.transports.File, {filename: './lnk95.log'});
generateMonthDays();
generateDayHours();


domoticz_linky.sh
#!/bin/sh
set -au

SCRIPT=$(readlink -f "$0")
BASE_DIR=$(dirname "${SCRIPT}")
export BASE_DIR
CFG_FILE="domoticz_linky.cfg"
LOG_FILE="domoticz_linky.log"
MONTH=0

if [ $# -eq 1 ]
then
    MODE="monthly"
    MONTH="$1"
else
    MODE="daily"
fi
echo "${MODE} mode" >> "${BASE_DIR}"/"${LOG_FILE}"

update_db () {
  if [ "$1" -eq 0 ]
  then
     PY_SCRIPT="linky_json.py"
  else
     PY_SCRIPT="linky_month.py"
  fi
  PY_SCRIPT="${BASE_DIR}"/"${PY_SCRIPT}"
  /usr/bin/python3 "${PY_SCRIPT}" $1 -o "${BASE_DIR}" >> "${BASE_DIR}"/"${LOG_FILE}" 2>&1
  if  [ $? -eq 0 ]; then
    BASE_DIR="${BASE_DIR}" /usr/bin/nodejs "${BASE_DIR}"/domoticz_linky.js "${DOMOTICZ_ID}" > "${BASE_DIR}"/req.sql
     cat "${BASE_DIR}"/req.sql | /usr/bin/sqlite3 "${HOME}"/domoticz/domoticz.db
  fi
}

# check configuration file
if [ -f "${BASE_DIR}"/"${CFG_FILE}" ]
then
  . "${BASE_DIR}"/"${CFG_FILE}"
  export LINKY_USERNAME
  export LINKY_PASSWORD
  export DOMOTICZ_ID
  update_db "${MONTH}"
else
    echo "Config file is missing ["${BASE_DIR}"/"${CFG_FILE}"]"
    exit 1
fi


linky_json.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
"""Generates energy consumption JSON files from Enedis (ERDF) consumption data
collected via their  website (API).
"""

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import datetime
import logging
import sys
import json
import linky
from dateutil.relativedelta import relativedelta

USERNAME = os.environ['LINKY_USERNAME']
PASSWORD = os.environ['LINKY_PASSWORD']
BASEDIR = os.environ['BASE_DIR']

# Generate y axis (consumption values) 
def generate_y_axis(res):
    y_values = []

    # Extract data points from the source dictionary into a list
    for ordre, datapoint in enumerate(res['graphe']['data']):
        value = datapoint['valeur']

        # Remove any invalid values
        # (they're error codes on the API side, but useless here)
        if value < 0:
            value = 0

        y_values.insert(ordre, value)

    return y_values

# Generate x axis (time values)  
def generate_x_axis(res, time_delta_unit, time_format, inc):
    x_values = []

    # Extract start date and parse it
    start_date_queried_str = res['graphe']['periode']['dateDebut']
    start_date_queried = datetime.datetime.strptime(start_date_queried_str, "%d/%m/%Y").date()

    # Calculate final start date using the "offset" attribute returned by the API
    kwargs = {}
    kwargs[time_delta_unit] = res['graphe']['decalage'] * inc
    start_date = start_date_queried - relativedelta(**kwargs)

    # Generate X axis time labels for every data point
    for ordre, _ in enumerate(res['graphe']['data']):
        kwargs = {}
        kwargs[time_delta_unit] = ordre * inc
        x_values.insert(ordre, (start_date + relativedelta(**kwargs)).strftime(time_format))

    return x_values

# Date formatting 
def dtostr(date):
    return date.strftime("%d/%m/%Y")


# Export the JSON file for half-hours power measure (for the last pas day)
def export_hours_values(res):
    hours_x_values = generate_x_axis(res, \
                                    'hours', "%H:%M", 0.5)
    hours_y_values = generate_y_axis(res)
    hours_values = []

    for i in range(0,len(hours_x_values)):
        hours_values.append({"time" : hours_x_values[i], "conso" : hours_y_values[i]})
    with open(BASEDIR+"/export_hours_values.json", 'w+') as outfile:
        json.dump(hours_values, outfile)

# Export the JSON file for daily consumption (for the past rolling 30 days)
def export_days_values(res):
    days_x_values = generate_x_axis(res, \
                                    'days', "%d %b", 1)
    days_y_values = generate_y_axis(res)
    days_values = []

    for i in range(0,len(days_x_values)):
        days_values.append({"time" : days_x_values[i], "conso" : days_y_values[i]})
    with open(BASEDIR+"/export_days_values.json", 'w+') as outfile:
        json.dump(days_values, outfile)

# Export the JSON file for monthly consumption (for the current year, starting 12 months from today)
def export_months_values(res):
    months_x_values = generate_x_axis(res, \
                                    'months', "%b", 1)
    months_y_values = generate_y_axis(res)
    months_values = []

    for i in range(0,len(months_x_values)):
        months_values.append({"time" : months_x_values[i], "conso" : months_y_values[i]})
    with open(BASEDIR+"/export_months_values.json", 'w+') as outfile:
        json.dump(months_values, outfile)

# Export the JSON file for yearly consumption
def export_years_values(res):
    years_x_values = generate_x_axis(res, \
                                    'years', "%Y", 1)
    years_y_values = generate_y_axis(res)
    years_values = []

    for i in range(0,len(years_x_values)):
        years_values.append({"time" : years_x_values[i], "conso" : years_y_values[i]})
    with open(BASEDIR+"/export_years_values.json", 'w+') as outfile:
        json.dump(years_values, outfile)




# Main script 
def main():
    logging.basicConfig(format='%(asctime)s %(message)s', level=logging.INFO)

    try:
        logging.info("logging in as %s...", USERNAME)
        token = linky.login(USERNAME, PASSWORD)
        logging.info("logged in successfully!")

        logging.info("retrieving data...")
        today = datetime.date.today()
        
        # Years
        res_year = linky.get_data_per_year(token)

        # 12 months ago - today
        res_month = linky.get_data_per_month(token, dtostr(today - relativedelta(months=11)), \
                                             dtostr(today))


        # One month ago - yesterday
        res_day = linky.get_data_per_day(token, dtostr(today - relativedelta(days=1, months=1)), \
                                         dtostr(today - relativedelta(days=1)))


        # Yesterday and the day before
        res_hour = linky.get_data_per_hour(token, dtostr(today - relativedelta(days=2)), \
                                           dtostr(today))
        

        logging.info("got data!")
############################################
		# Export of the JSON files, with exception handling as Enedis website is not robust and return empty data often
        try:
            export_hours_values(res_hour)
        except Exception as exc:
        	# logging.info("hours values non exported")
            logging.error(exc)

        try:
            export_days_values(res_day)
        except Exception:
            logging.info("days values non exported")
            sys.exit(70)

        try:
            export_months_values(res_month)
        except Exception:
            logging.info("months values non exported")

        try:
            export_years_values(res_year)
        except Exception:
        	logging.info("years values non exported")

############################################
 
    except linky.LinkyLoginException as exc:
        logging.error(exc)
        sys.exit(1)



if __name__ == "__main__":
    main()
